### DISCUSSION TITLE
Rendering engine in terminal

### DISCUSSION THREAD

**Discussion by psi4j** [neovide/neovide#2616] [Ideas]
Is it possible to extend the rendering of the cursor animations to the terminal buffers inside Neovim and or use Neovide to run terminal emulators and or TMUX directly with the animations? If I could run my entire setup inside Neovide I would. I'm assuming there are some large hurdles to cross here and that must be why it hasn't already been done, but I'm wondering if this is something we can expect to be implemented in the foreseeable future. Thank you guys for your hard work on Neovide.

**Comment by typecasto** [comment]
You could make a custom neovim config that opens a terminal on startup and unbind all the commands that exit insert mode, if you really wanted to.

**Reply by psi4j** [reply]
Thank you for your response and for the tip. Is there currently a way have Neovide take over the rendering of the cursor inside of the Neovim terminals and apply its cursor animations (movements, the new blinking effect, etc)? I haven't been able to get it to apply its cursor animations, though it applies all of my other settings, including p10k, other zsh settings in my .zshrc, and the colors given to the terminal using my lua config. There's something that I'm probably missing here. Maybe there's already a way to do what I'm talking about.

**Comment by xzbdmw** [comment]
Neovim draw builtin terminal themselves and don't report cursor position to neovide, you can see the cursor in terminal is a block(it is a fake cursor just has background highlighted)
see 
https://github.com/neovide/neovide/discussions/2495
https://github.com/neovide/neovide/issues/2445

**Reply by psi4j** [reply]
Oh, I think I understand. So are you saying it's not currently possible and maybe not something that'll be included anytime soon?

**Reply by xzbdmw** [reply]
Right,I’m quite annoyed by that block cursor, in theory you can hack neovide to show real cursor if it is about to render TermCursor hah.See https://github.com/neovim/neovim/issues/3681

**Reply by psi4j** [reply]
> Right,I’m quite annoyed by that block cursor, in theory you can hack neovide to show real cursor if it is about to render TermCursor hah.See https://github.com/neovim/neovim/issues/3681

Thank you for clarifying. More reason for me to work on my Rust skills to maybe eventually take a shot at this myself.

**Reply by xzbdmw** [reply]
In fact it already works, by deleting [this check](https://github.com/neovide/neovide/blob/74170f4884322ea8f236b66308feba1e2b3e1912/src/renderer/cursor_renderer/mod.rs#L284), you will see neovide cursor in terminal.

**Reply by psi4j** [reply]
> In fact it already works, by deleting [this check](https://github.com/neovide/neovide/blob/74170f4884322ea8f236b66308feba1e2b3e1912/src/renderer/cursor_renderer/mod.rs#L284), you will see neovide cursor in terminal.

I will look into this when a get a chance. I'll have to rebuild Neovide though, huh? I'm not familiar with the code and if this could have any additional side effects. It'd be nice to have this as a battle-tested option out of the box. The main selling point for Neovide is its visually-guiding cursor animations, so it'd be optimal to have it everywhere with the program.

**Reply by xzbdmw** [reply]
I'm not sure the side effects, neovim reports cursor position in terminal but with enable=false, that's why neovide chooses not to render. Before yesterday I thought neovim does not report cursor position at all 😄 , anyway, seeing a vertical bar in terminal is such a relief.
I think this will still be an upstream issue, it's just a simple hack. But build neovide is very easy.

**Reply by psi4j** [reply]
> I'm not sure the side effects, neovim reports cursor position in terminal but with enable=false, that's why neovide chooses not to render. Before yesterday I thought neovim does not report cursor position at all 😄 , anyway, seeing a vertical bar in terminal is such a relief. I think this will still be an upstream issue, it's just a simple hack. But build neovide is very easy.

I got Neovide to render the cursor inside the terminal by doing what you recommended. Do you know if there is a way to simply set Neovim to report the cursor location? You said Neovim reports cursor position in the terminal, but with enable=false. What do you mean by this?

A side effect I noticed is that now when I use the yazi terminal file manager through the yazi plugin for Neovim, it renders the cursor at the end of the file preview, which is a little odd, but I can deal with it until I figure out how to stop this.

**Reply by xzbdmw** [reply]
> > I'm not sure the side effects, neovim reports cursor position in terminal but with enable=false, that's why neovide chooses not to render. Before yesterday I thought neovim does not report cursor position at all 😄 , anyway, seeing a vertical bar in terminal is such a relief. I think this will still be an upstream issue, it's just a simple hack. But build neovide is very easy.
> 
> I got Neovide to render the cursor inside the terminal by doing what you recommended. Do you know if there is a way to simply set Neovim to report the cursor location? You said Neovim reports cursor position in the terminal, but with enable=false. What do you mean by this?
> 
> A side effect I noticed is that now when I use the yazi terminal file manager through the yazi plugin for Neovim, it renders the cursor at the end of the file preview, which is a little odd, but I can deal with it until I figure out how to stop this.

yeah, lazygit also show the cursor at right bottom corner, it’s the price to pay maybe, not a big deal haha.

enabled is a field of msgpackrpc protol message of cursor_goto. see https://github.com/neovim/neovim/pull/26835

**Comment by TRPB** [comment]
Neovide running a terminal directly with animations and smooth scrolling would instantly make it the best terminal available once the transparency weirdness is solved! 

Would be nice to be able to launch zsh, rather than neovim->terminal->zsh but I'm probably making it sound way easier than it is.
